File: test-rate-limiter.ts
import ccxt from "ccxt";

const exchange = new ccxt.binance ({
    enableRateLimit: true,
    // rateLimiterAlgorithm: "leakyBucket",
    rateLimiterAlgorithm: "rollingWindow",
    maxLimiterRequests: 5000,
    timeout: 90000,
});

async function fetchTickersParallel(times: number, symbol: string): Promise<void> {
    const startTime = Date.now();
    const requests = [...Array(times)].map(async (_, index) => {
        try {
            const result = await exchange.fetchTicker("BTC/USDT");
            console.log(`Call ${index + 1} completed in ${Date.now() - startTime} ms`);
            return result;
        } catch (error) {
            console.error(`Call ${index + 1} failed: ${error}`);
        }
    });
    await Promise.all(requests);
}

const numRequests = 1000;
fetchTickersParallel(numRequests, "BTC/USDT").catch(console.error);

File: test-rate-limiter.py
import asyncio
import time
import sys
sys.path.insert(0, '/Users/caoilainn/Documents/ccxt/python')
import ccxt.async_support as ccxt


async def fetch_tickers_parallel(times, symbol):
    exchange = ccxt.binance({
        'enableRateLimit': True,
        # 'rateLimiterAlgorithm': 'leakyBucket',
        'rateLimiterAlgorithm': 'rollingWindow',
        'maxLimiterRequests': 1000,
    })
    try:
        start_time = time.time()
        tasks = []
        for i in range(times):
            tasks.append(fetch_ticker(exchange, symbol, i, start_time))
        await asyncio.gather(*tasks)
    finally:
        await exchange.close()


async def fetch_ticker(exchange, symbol, index, start_time):
    try:
        result = await exchange.fetch_ticker(symbol)
        elapsed = (time.time() - start_time) * 1000
        print(f"Call {index + 1} completed in {elapsed:.2f} ms")
        return result
    except Exception as e:
        print(f"Call {index + 1} failed: {e}")


num_requests = 5000
asyncio.run(fetch_tickers_parallel(num_requests, 'BTC/USDT'))
